Extend Submission Requirement for Arbitrary Logic (#19)
decentralized-identity/presentation-exchange: spec/spec.md

  `input_descriptor` array are required for submission. The composition of
  values under this property are described in the [`Input
  Descriptors`](#input-descriptors) section below.

### Submission Requirement

_Presentation Definitions_ ****MAY**** include a _Submission Requirement_,
which defines combinations of inputs that comply with an Issuer or Verifier
flow, such as credential issuance. The _Submission Requirement_ can be
interpreted by a User Agent as instructions on how to present requirement
optionality to the user and submit a valid combination of inputs back via a
_Proof Submission_ object. The following section defines the format for the
_Submission Requirement_ object and the selection syntax used to specify valid
input combinations.

::: example Submission Requirement
```json
  "submission_requirement": {
    name: "Credential issuance requirements",
    purpose: "Verify banking, employment, and citizenship information.",
    rule: "all",
    from: [
      {
        "name": "Banking Information",
        "purpose": "We need to know if you have an established banking history.",
        "rule": "pick",
        "count": 1,

>         "from": ["A"]

Reviewer: ```suggestion
        "from": "A"
```
Should this match the example above?

Author: yes!

      },
      {
        "name": "Employment Information",
        "purpose": "We need to know that you are currently employed.",
        "rule": "all",

>         "from": ["B"]

Reviewer: ```suggestion
        "from": "B"
```

      }
      {
        "name": "Citizenship Information",
        "rule": "pick",
        "count": 1,

>         "from": ["C"]

Reviewer: ```suggestion
        "from": "C"
```

      }
    ]
  },
```
:::

#### Submission Requirement Objects

_Submission Requirement Objects_ describe what combinations of inputs will
satisfy Verifier requires for evaluation in a subsequent [Presentation
Submission](#presentation-submission). _Requirement Objects_ are JSON objects
constructed as follows:

1. The object ****MUST**** contain a `rule` property, and its value
   ****MUST**** be a string matching one of the [Submission Requirement
   Rules](#submission-requirement-rules) values listed in the section below.
2. The _Submission Requirement_ object ****MUST**** contain a `from` property,
   and its value ****MUST**** be one of

	(a) A `group` string matching one or more of the _Input Descriptor_
	objects in the `input_descriptors`

	(b) An array of nested _Submission Requirement_ objects.

3. The object ****MAY**** contain a `name` property, and if present, its value

>    ****MUST**** be a string of the creator's choosing, which ****MAY**** be

Reviewer: ```suggestion
   ****MUST**** be a string which ****MAY**** be
```

   used by a consuming User Agent to display the general name of the
   requirement set to a user.

> 4. The object ****MAY**** contain a `purpose` property, and if present, its

Reviewer: ```suggestion
4. The object ****MAY**** contain a `purpose` property and, if present, its
```

the _Presentation Definition_ to submit all members of the matching `group`
string. In the following example, the `from` property contains a `group`
string to require a single member of group `"B"`:

::: example Submission Requirement, pick, group
  ```json
  "submission_requirement": {
    "name": "Picking one member from group B",
    "purpose": "We only need one",
    "rule": "pick",
    "count": 1,
    "from": "B"
  }
  ```
:::

If the `from` property contains an array of nested _Submission Requirement_
objects, it directs the consumer of the _Presentation Definition_ to submit
members such that the number of satisfied _Submission Requirement_ objects is
exactly `count`. In the following example, the `from` property contains an
array of nested _Submission Requirement_ objects to require either all members
from group `"A"` or two members from group `"B"`:

::: example Submission Requirement, pick, nested
  ```json
  "submission_requirement": {
    "name": "Either all from group A or two from group B",
    "purpose": "Either way works",
    "rule": "pick",
    "count": 1,
    "from": [
      {"rule": "all",  "from": "A"},
      {"rule": "pick", "from": "B", "count": 2},
    ]

>   }

Reviewer: I like this use of nesting which allows a `pick` rule to operate as a logical `or`

Author: ![image](https://user-images.githubusercontent.com/858687/84558168-c8472d80-acfe-11ea-855a-907da6718cf8.png)

  objects within the `input_descriptors` array.
- A _Submission Requirement_ is satisfied with respect to a _Presentation
  Submission_, as defined by the following algorithm:
    - If `from` is a string, then the rule refers to a `group` string defined
	within the _Input Descriptors_, and:

      - If `rule` is `"all"`, then this _Submission Requirement_ is satisfied
	if and only if the _Presentation Submission_ entries match all of the
	`group` string's members.

      - If `rule` is `"pick"`, then this _Submission Requirement_ is satisfied
	if and only if the number of _Presentation Submission_ entries matching
	`group` string's members is exactly `count`.

    - If `from` is an array of _Submission Requirement_ objects, then the
	rule refers to nested requirements, and:

      - If `rule is `"all"`, then this _Submission Requirement_ is satisfied if
	and only if all nested _Submission Requirement_ objects are satisfied.
      - If `rule` is `"pick"`, then this _Submission Requirement_ is satisfied
	if and only if the number of satisfied nested _Submission Requirement_
	objects is exactly `count`.

### Input Descriptors

> _Input Descriptors_ are objects used to describe the proofing inputs a Verifier requires of a Subject before they will proceed with an interaction. _Input Descriptor Objects_ contain a schema URI that links to the schema if the required input data, constrains on data values, and an explanation of why a certain set or item of data is being requested:

Reviewer: ```suggestion
_Input Descriptors_ are objects used to describe the proofing inputs a Verifier requires
of a Subject before they will proceed with an interaction. _Input Descriptor Objects_
contain a schema URI that links to the schema of the required input data, constraints
on data values, and an explanation why a certain item or set of data is being requested:
```

#### Input Descriptor Objects

_Input Descriptors_ are objects that describe what type of input data/credential, or sub-fields thereof, is required for submission to the Verifier. _Input Descriptor Objects_ are composed as follows:

>   - The object ****MUST**** contain an `id` property, and if present, its value ****MUST**** be a unique identifying string that does not conflict with the `id` of another _Input Descriptor_ in the same _Presentation Definition_ object.

Reviewer: ```suggestion
  - The object ****MUST**** contain an `id` property. The value of the `id` property ****MUST**** be a unique identifying string that does not conflict with the `id` of another _Input Descriptor_ in the same _Presentation Definition_ object.
```

Reviewer: I know this isn't a comment on a change you've recommended, but it needs fixing.
